### ISSUE TITLE
rsync issues with deployment

### ISSUE THREAD

**Issue by bottoy** [microsoft/Oryx#437]
Hi,

We encountered today an issue with the deployment of a new version of our Django.
When deploying on an App Service via git, we got following error:

`remote: rsync: failed to set times on "/home/site/wwwroot/antenv/lib64": Operation not permitted (1)`   

Each time we re-deploy the same issue persist. 

```
remote: Deploy Async        
remote: Updating branch 'master'.        
remote: Updating submodules.        
remote: Preparing deployment for commit id '05e2ced2f5'.        
remote: Repository path is /home/site/repository        
remote: Running oryx build...        
remote: Build orchestrated by Microsoft Oryx, https://github.com/Microsoft/Oryx        
remote: You can report issues at https://github.com/Microsoft/Oryx/issues        
remote: 
remote: Oryx Version      : 0.2.20191105.2, Commit: 67e159d71419415435cb5d10c05a0f0758ee8809, ReleaseTagName: 20191105.2        
remote: Build Operation ID: |9grRtIzpvr8=.64c3971e_        
remote: Repository Commit : 05e2ced2f51e57cf0b6372a8a5de3bc4c4baf995        
remote: 
remote: 
remote: Source directory     : /home/site/repository        
remote: Destination directory: /home/site/wwwroot        
remote: 
remote: Python Version: /opt/python/3.7.5/bin/python3        
remote: Python Virtual Environment: antenv        
remote: Creating virtual environment ...        
remote: Activating virtual environment ...        
remote: 
remote: Upgrading pip...        
remote: Requirement already up-to-date: pip in ./antenv/lib/python3.7/site-packages (19.3.1)        
remote: Done in 5 sec(s).        
remote: Running pip install...        
remote: [12:53:00+0000] Requirement already satisfied: adal==1.2.1 in ./antenv/lib/python3.7/site-packages (from -r requirements.txt (line 1)) (1.2.1)        
remote: [12:53:00+0000] Requirement already satisfied: applicationinsights==0.11.7 in ./antenv/lib/python3.7/site-packages (from -r requirements.txt (line 2)) (0.11.7)        
...
ruamel.yaml>=0.15.34->drf-yasg==1.16.1->-r requirements.txt (line 35)) (0.2.0)        
remote: Done running pip install.        
remote: 
remote: Copying files to destination directory '/home/site/wwwroot'...        
remote: rsync: failed to set times on "/home/site/wwwroot/antenv/lib64": Operation not permitted (1)        
remote: rsync: rename "/home/site/wwwroot/antenv/lib/python3.7/site-packages/._brotli.cpython-37m-x86_64-linux-gnu.so.pf53bJ" -> "antenv/lib/python3.7/site-packages/_brotli.cpython-37m-x86_64-linux-gnu.so": No such file or directory (2)        
remote: rsync: rename "/home/site/wwwroot/antenv/lib/python3.7/site-packages/._cffi_backend.cpython-37m-x86_64-linux-gnu.so.mWM9Cl" -> "antenv/lib/python3.7/site-packages/_cffi_backend.cpython-37m-x86_64-linux-gnu.so": No such file or directory (2)        
remote: rsync: rename "/home/site/wwwroot/antenv/lib/python3.7/site-packages/._ruamel_yaml.cpython-37m-x86_64-linux-gnu.so.dGqgcY" -> "antenv/lib/python3.7/site-packages/_ruamel_yaml.cpython-37m-x86_64-linux-gnu.so": No such file or directory (2)        
remote: rsync: rename "/home/site/wwwroot/antenv/lib/python3.7/site-packages/.libs_cffi_backend/.libffi-d78936b1.so.6.0.4.sAAfXA" -> "antenv/lib/python3.7/site-packages/.libs_cffi_backend/libffi-d78936b1.so.6.0.4": No such file or directory (2)        
remote: rsync: rename "/home/site/wwwroot/antenv/lib/python3.7/site-packages/Cryptodome/Cipher/._Salsa20.cpython-37m-x86_64-linux-gnu.so.KelzTd" -> "antenv/lib/python3.7/site-packages/Cryptodome/Cipher/_Salsa20.cpython-37m-x86_64-linux-gnu.so": No such file or directory (2)       
```

Any clue what causes these errors?

Kind regards,
Bram

**Comment by kichalla** [comment]
cc @sanchitmehta

**Comment by sdementen** [comment]
hello, this bug is really painful for us.

We wanted to fix a bug in our app and deploy it and now we are facing theses bugs in the build&deploy script on Azure Webapp forbidding us to deploy this hotfix. 
We did not get any notifications about potential issues regarding a next deployment so we are totally caught by surprise. No fun :-(

**Comment by arroyc** [comment]
@sdementen , we are really sorry for the bad experience, we are working on the fix to be deployed (as far as last update), I will follow up as soon as I have some good news. Once again we apologize for the poor experience, and at the same time really appreciate you reporting these bugs and issues to improve our product. Looking forward to your support as always

**Comment by bottoy** [comment]
It would be nice that via an environment variable we can select the version of the script to run. If not set it will run the latest version, if set, it will run that version.  This will allow a roll-back scenario ...
Condition off course is that redundancy in the script is possible....

**Comment by arroyc** [comment]
@bottoy .. thank you for suggestion, there is deployment freeze due to holiday season, still we are trying our best to deploy this fx to production.
Regarding roll back scenario, in general app settings you can try setting this ENABLE_ORYX_BUILD to false, let me know if that works for now

**Comment by bottoy** [comment]
@arroyc IT works with this flag, we can deploy and the website starts up via kudu.

Could you give a sign when we can use the Oryx build-script again? Thx

**Comment by arroyc** [comment]
I will certainly do that .. thanks @bottoy

**Comment by CobraClutch** [comment]
I fought with this for way too many hours before finally disabling Oryx altogether. Once I set ENABLE_ORYX_BUILD to false, my node_modules folder magically appeared. Could it be something to do with my package.json being in a subdirectory?

**Comment by davjfish** [comment]
We are running into a similar issue when trying to deploy a python linux web app (django) to a web app service through a release pipeline, i.e. zip deployment.

When starting with a clean sandbox / deployment slot, the zip deployment is successful. Upon any subsequent release, deployment fails with the following output:

```
Copying files to destination directory '/home/site/wwwroot'...
rsync: rename "/home/site/wwwroot/.azure_oauth_settings.yml.t9QlcB" -> "azure_oauth_settings.yml": No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1196) [sender=3.1.2]
rsync: rename "/home/site/wwwroot/.azure_oauth_settings.yml.t9QlcB" -> "azure_oauth_settings.yml": No such file or directory (2)\nrsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1196) [sender=3.1.2]\n/opt/Kudu/Scripts/starter.sh oryx build /tmp/zipdeploy/extracted -o /home/site/wwwroot --platform python --platform-version 3.7 -i /tmp/8d7b47b2c4d1ee1 -p compress_virtualenv=tar-gz -p virtualenv_name=antenv --log-file /tmp/build-debug.log 
##[error]Failed to deploy web package to App Service.
##[error]Error: Package deployment using ZIP Deploy failed. Refer logs for more details.
##[warning]Error: Failed to update deployment history. Error: Bad Request (CODE: 400)
``` 

The file that it is complaining about `azure_oauth_settings.yml` is not a part of the git repo. It is being built as a part of the CI pipeline and sent to the release pipeline as an artifact. 

Here are some things that I find confusing: 
1) this is not the only file that is in the artifact but not part of the git repo. 
2) after the first deployment, if I SSH to the web app sandbox, I am able to interact with the above-mentioned file, e.g., `cat /home/site/wwwroot/azure_oauth_settings.yml`. After a failed release, if I try to repeat the above command, I get the same error that we are getting above in the rsync operation (i.e., `No such file or directory`), even though the file seems to be present in the directory:

![image](https://user-images.githubusercontent.com/34521117/74756136-7e67a280-524a-11ea-9456-749077ed6ff4.png)



I also noticed a warning about a failure to update deployment history. I am not sure if this would be related to the problem or not.

**Comment by arroyc** [comment]
hey guys, app service recently updated toolings across their regions .. can you please retry and report us back in case you are facing similar issues? 
Also, AppService team now supports 'running from package' which is supposed to fix issues like this. Could you try with this along with setting ENABLE_ORYX_BUILD to true and let us know if you are still seeing this issue?
https://docs.microsoft.com/bs-latn-ba/azure/app-service/deploy-run-package

**Comment by davjfish** [comment]
We ended up implementing a work-around, sorry.. the problematic code is long-gone... Thanks anyways for the followup

**Comment by arroyc** [comment]
@davjfish  . .thanks for responding ..  would you like to share the work-around?  In that case, I will be closing this issue, please feel free to open a new one

**Comment by phdowling** [comment]
Hi, currently seeing this issue out of the blue on a python app service in Azure. Anything I can do except nuke the app service and start fresh?

**Comment by davjfish** [comment]
@phdowling I don't remember exactly how we worked around the issue, but here are some vague memories in case they might be helpful. The bug seemed to be triggered by _the way_ the application was trying to connect to the file. I think the package we were using to ingest the yaml file (possibly pyYAML) was locking it down and this is what was preventing subsequent deployments. In our case, we were able to switch from a YAML to .env file and read in the data using the `python-decouple` package. This seemed to resolve the issue for us.
